### PR TITLE
Add BKTClient

### PR DESCRIPTION
**Pull request by jinSasaki**
Please review after #86 

Add BKTClient and tests

### CODE REVIEW

File: Bucketeer2/Sources/Internal/Model/Evaluation.swift [hunk 1]
```diff
@@ -10,3 +10,32 @@ struct Evaluation: Hashable, Codable {
...
+        default:
+            anyValue = value
+        }
+        guard let typedValue = anyValue as? T else {
+            logger?.debug(message: "getVariation returns null reason: failed to cast")
+            return defaultValue
+        }
+        return typedValue
+    }
+}
```
Comment (jinSasaki) [inline Bucketeer2/Sources/Internal/Model/Evaluation.swift:41]: Getting as a specified type from text variation value.

File: Bucketeer2/Sources/Internal/Scheduler/TaskScheduler.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    let component: Component
+    let dispatchQueue: DispatchQueue
+
+    private lazy var foregroundSchedulers: [ScheduledTask] = [
+        EvaluationForegroundTask(component: component, queue: dispatchQueue),
+        EventForegroundTask(component: component, queue: dispatchQueue)
+    ]
+
+    private lazy var backgroundSchedulers: [ScheduledTask] = [
+        // TODO: Add background scheduler
```
Comment (jinSasaki) [inline Bucketeer2/Sources/Internal/Scheduler/TaskScheduler.swift:13]: It is hard for sdk scheduling in the background. I'm looking for that approach.

File: Bucketeer2Tests/Mock/MockEventInteractor.swift [hunk 3]
```diff
@@ -24,13 +24,22 @@ final class MockEventInteractor: EventInteractor {
...
+    func trackEvaluationEvent(featureTag: String, user: User, evaluation: Evaluation) throws {
+
+    }
+    func trackDefaultEvaluationEvent(featureTag: String, user: User, featureId: String) throws {
+
+    }
+    func trackGoalEvent(featureTag: String, user: User, goalId: String, value: Double) throws {
+
+    }
+    func trackFetchEvaluationsSuccess(featureTag: String, seconds: Int64, sizeByte: Int64) throws {
```
Comment (jinSasaki) [inline Bucketeer2Tests/Mock/MockEventInteractor.swift:36]: I will add a handler if it is needed.

File: Bucketeer2/Sources/Internal/DI/Component.swift [hunk 4]
```diff
@@ -9,29 +9,17 @@ protocol Component: AnyObject {
...
-
-    var config: BKTConfig {
-        dataModule.config
-    }
-
-    var userHolder: UserHolder {
-        dataModule.userHolder
-    }
-
-    var evaluationInteractor: EvaluationInteractor {
```
Comment (jinSasaki) [inline Bucketeer2/Sources/Internal/DI/Component.swift]: The computed property pattern has a bug that the cache data was not shared between instances.

File: Bucketeer2/Sources/Internal/Model/Evaluation.swift [hunk 5]
```diff
@@ -10,3 +10,31 @@ struct Evaluation: Hashable, Codable {
...
+        case is Int64:
+            anyValue = Int64(value)
+        case is Float:
+            anyValue = Float(value)
+        case is Double:
+            anyValue = Double(value)
+        case is Bool:
+            anyValue = Bool(value)
+        default:
+            anyValue = value
```
Comment (cre8ivejp) [inline Bucketeer2/Sources/Internal/Model/Evaluation.swift:32]: Is there a way to convert JSON objects in Swift to validate them?
By checking if it is a dictionary or not?

Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Model/Evaluation.swift:32]: Yes, it is possible.
I can also convert it to `[String: AnyHashable]` and check it.
Can I change [`BKTClient jsonVariation`](https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/files/144e2a02474e5861ada9727fafb82a4683956c4c..4a221778df30bcdba879c7e6a7e2d8c5d69e5bf7#diff-ecc95802cf8595971f48fdbb7da52ef1181e037b9dd20132058a84345117f4e4R8) `Any` to `[String: AnyHashable])` if that is the case?

File: Bucketeer2/Sources/Public/BKTClient.swift [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+extension BKTClient {
+    public static func initialize(config: BKTConfig, user: BKTUser, timeoutMillis: Int64 = 5000, completion: ((BKTError?) -> Void)?) {
+        guard BKTClientImpl.default == nil else {
+            config.logger?.warn(message: "BKTClient is already initialized. not sure if initial fetch has been finished")
+            return
+        }
+        do {
+            let dispatchQueue = DispatchQueue(label: "jp.co.cyberagent.bucketeer.taskQueue")
```
[the remark this answers is not in the document]
Comment (jinSasaki) [reply Bucketeer2/Sources/Public/BKTClient.swift]: changed https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/commits/90393647645e3e0794545e21649b91b577f03a3c

File: Bucketeer2/Sources/Public/BKTClient.swift [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    func updateUserAttributes(attributes: [String: String])
+    func fetchEvaluations(timeoutMillis: Int64?, completion: ((BKTError?) -> Void)?)
+    func flush(completion: ((BKTError?) -> Void)?)
+    func evaluationDetails(featureId: String, completion: ((BKTEvaluation?) -> Void)?)
+}
+
+extension BKTClient {
+    public static func initialize(config: BKTConfig, user: BKTUser, timeoutMillis: Int64 = 5000, completion: ((BKTError?) -> Void)?) {
+        guard BKTClientImpl.default == nil else {
+            config.logger?.warn(message: "BKTClient is already initialized. not sure if initial fetch has been finished")
```
[the remark this answers is not in the document]
Comment (jinSasaki) [reply Bucketeer2/Sources/Public/BKTClient.swift]: fixed https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/commits/90393647645e3e0794545e21649b91b577f03a3c

File: Bucketeer2/Sources/Internal/Model/Evaluation.swift (hunk 5 above)
Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Model/Evaluation.swift]: As a trial, what do you think of the change? https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/commits/68457f0ac275c5b57b521299716249a0a57c6553

File: Bucketeer2/Sources/Internal/Model/Evaluation.swift [hunk 8]
```diff
@@ -29,7 +29,12 @@ extension Evaluation {
         case is Bool:
             anyValue = Bool(value)
         default:
-            anyValue = value
+            let data = value.data(using: .utf8) ?? Data()
```
[the remark this answers is not in the document]
Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Model/Evaluation.swift:32]: Added JSON case
https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/commits/138ab2f2428b7ac1cd73f553939335c721259b29

File: Bucketeer2/Sources/Internal/Model/Evaluation.swift [hunk 9]
```diff
@@ -10,3 +10,35 @@ struct Evaluation: Hashable, Codable {
...
+        let value = self.variation_value
+        let anyValue: Any?
+        switch defaultValue {
+        case is String:
+            anyValue = value
+        case is Int:
+            anyValue = Int(value)
+        case is Int64:
+            anyValue = Int64(value)
+        case is Float:
```
Comment (cre8ivejp) [inline Bucketeer2/Sources/Internal/Model/Evaluation.swift:25]: I just noticed we don't need the Float and Int64 cases.

https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/files#diff-ecc95802cf8595971f48fdbb7da52ef1181e037b9dd20132058a84345117f4e4R108-R126

Comment (jinSasaki) [reply Bucketeer2/Sources/Internal/Model/Evaluation.swift]: You are right! Deleted.
https://github.com/ca-dp/bucketeer-ios-sdk/pull/87/commits/bcdf73d0d9a296c2abc5a685d086c1cd319997a0

Comment (cre8ivejp) [reply Bucketeer2/Sources/Internal/Model/Evaluation.swift]: Thank you!

**Review by cre8ivejp — APPROVED** [review]
👍

**Pull request merged**